### PR TITLE
3233 Add CumulativeAverage metric

### LINKED ISSUE #3233: Add AverageMeter metric handler
**Is your feature request related to a problem? Please describe.**
It would be useful to have some AverageMeter metric or handler to continuously computing the average value during training, draft idea code:
```py
class AverageMeter(object):
    """Computes and stores the average and current value"""
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        if n > 0:
            self.val = val
            self.sum += val * n

            self.count += n
            self.avg = self.sum / self.count if self.count > 0 else self.sum
```

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #3233 .

### Description
This PR added the `CumulativeAverage` metric to continuously compute average on some input data, it can support distributed data-parallel and multi-classes.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/metrics/cumulative_average.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import torch
+
+from monai.metrics.utils import do_metric_reduction
+from monai.utils import MetricReduction
+
+from .metric import CumulativeIterationMetric
+
+
+class CumulativeAverage(CumulativeIterationMetric):
```
Comment (wyli) [inline monai/metrics/cumulative_average.py:22]: I think it's easier to follow if it inherits `Cumulative`, it doesn't have to be metrics...

Comment (Nic-Ma) [reply monai/metrics/cumulative_average.py]: Thanks for your suggestion.
That's good idea, I updated in the latest commit.
Could you please help review it again?

Thanks.

**Review by myron — APPROVED** [review]
looks good

File: monai/metrics/cumulative_average.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        self.sum = None
+        self.not_nans = None
+
+    def aggregate(self):  # type: ignore
+        """
+        Sync data from all the ranks and compute the average value with previous sum value.
+
+        """
+        data = self.get_buffer()
+        if not isinstance(data, torch.Tensor):
```
Comment (wyli) [inline monai/metrics/cumulative_average.py:60]: can this class be generic, not limited to tensor operations?

File: monai/metrics/cumulative_average.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            raise ValueError("the data to aggregate must be PyTorch Tensor.")
+
+        # do metric reduction
+        f, not_nans = do_metric_reduction(data, reduction=MetricReduction.SUM_BATCH)
```
Comment (wyli) [inline monai/metrics/cumulative_average.py:64]: this is the only line that the class is related to "metric", do we really need this in this class?

File: monai/metrics/cumulative_average.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        # clear the buffer for next update
+        super().reset()
```
Comment (wyli) [inline monai/metrics/cumulative_average.py:66]: why reset the buffer here? in my opinion `self.sum` and `self.non_nans` could be local variables and they could always be computed from `self._buffers`?

**Review by wyli — COMMENTED** [review]
thanks, please see comments inline

File: monai/metrics/cumulative_average.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/metrics/cumulative_average.py:66]: Hi @wyli ,

Actually, @myron and I also discussed about this logic today, it looks more clear to always compute from `self.buffers`, but it's better to clear the buffer after every aggregation and only keep the `sum` & `not_nans` for the next computation to **save memory** and **performance**, otherwise, we need to repeatedly compute the average value from all the buffer when aggregating.
What do you think?

Thanks.

File: monai/metrics/cumulative_average.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/metrics/cumulative_average.py:64]: Hi @wyli ,

Here we leveraged the reduction utility to compute SUM on not_nans, do you mean to compute here directly?
```py
not_nans = (~(torch.isnan(data))).float().sum(dim=0)
f = data.sum(dim=0)  # the batch sum
```

Thanks.

File: monai/metrics/cumulative_average.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/metrics/cumulative_average.py:60]: OK, I think maybe this line is not necessary, to support multi-GPUs sync, the data in the base class `Cumulative` is already must be `torch.Tensor`.
I added this line because the following `SUM` logic is based on torch.Tensor APIs.

Thanks.

File: monai/metrics/cumulative_average.py (hunk 4 above)
Comment (wyli) [reply monai/metrics/cumulative_average.py:66]: To compute the running average perhaps this class should override the append and extend?

Comment (Nic-Ma) [reply monai/metrics/cumulative_average.py:66]: Hi @wyli ,

OK, that's an interesting idea, let me think about it.
The only obvious difference is that we need to sync across ranks before computation.

Thanks.

Comment (myron) [reply monai/metrics/cumulative_average.py:66]: @wyli I agree with Nic about clearing memory. We need to do aggregation for logging frequently (every iteration), if don't clear, then the GPU memory array will grow constantly and require sync across gpu on a larger size every time.
@Nic-Ma if it's easy to chage append (you can try), but I need it for MIL the sooner the better

File: monai/metrics/cumulative_average.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/metrics/cumulative_average.py]: removed in the latest commit.

Thanks.

File: monai/metrics/cumulative_average.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/metrics/cumulative_average.py]: Updated in the latest commit and added test for numpy data.

Thanks.

File: monai/metrics/cumulative_average.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/metrics/cumulative_average.py:68]: Hi @wyli ,

I updated the PR based on the online discussion, could you please help review it again?

Thanks in advance.

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me.
